Make TranscodingStream resilient against partial reads

GrabYourPitchforks opened the pull request:
Resolves https://github.com/dotnet/runtime/issues/49127.

If the call to `Encoder.GetBytes` return __0__, this doesn't necessarily mean that EOF has been reached. It could be that we did read data from the inner stream, but it wasn't enough data for us to make forward progress. If we let this __0__ return value propagate up to our caller, we'll incorrectly signal to them that we saw EOF.

The fix is to add a defensive loop, only breaking out of the loop when: (a) we really do see EOF from the inner stream; _or_ (b) the call to `GetBytes` was able to make non-zero forward progress.

/cc @tarekgh as FYI, and thanks @eiriktsarpalis for your minimal repro in the linked issue!

Note to reviewers: recommend reviewing with [_ignore whitespace changes_](https://github.com/dotnet/runtime/pull/50886/files?diff=split&w=1).

GrabYourPitchforks committed:
Make TranscodingStream resilient against partial reads

eiriktsarpalis commented on src/libraries/System.Text.Encoding/tests/Encoding/TranscodingStreamTests.cs:
```diff
@@ -336,10 +336,10 @@ private void RunReadTest(Func<Stream, MemoryStream, int> callback)
 
             Assert.Equal(-1, transcodingStream.ReadByte()); // should've reached EOF
 
-            // Now put some invalid data into the inner stream as EOF.
```
Could you also add a test that specifically validates the bug as reported? Thanks.

GrabYourPitchforks replied:
That specific scenario test does exist. See lines 359, 549 of this file.

GrabYourPitchforks committed:
Add extended async tests to TranscodingStream

GrabYourPitchforks commented:
@eiriktsarpalis See if the latest iteration addresses your concerns. I added some tests specifically to track the returned tasks in the async invalid and partial read cases.

eiriktsarpalis commented on src/libraries/System.Text.Encoding/tests/Encoding/TranscodingStreamTests.cs:
```diff
@@ -452,6 +470,56 @@ public async Task ReadAsync_Memory(int bufferLength)
+            readBuffer = new byte[1024];
+            comms.WriteBytes(new byte[] { 0xC2 });
+
+            task = transcodingStream.ReadAsync(readBuffer.AsMemory());
```
Would it make sense to add a test that validates behaviour when passing a cancellation token in this scenario?

GrabYourPitchforks replied:
That's already part of the main `ReadAsync` test logic. See below.

https://github.com/dotnet/runtime/blob/b1a417875e68e1961c163c1f0611549c2c2e04f6/src/libraries/System.Text.Encoding/tests/Encoding/TranscodingStreamTests.cs#L487-L504

GrabYourPitchforks commented:
Failing test is known issue https://github.com/dotnet/runtime/issues/50300.
